geps/gep-1762.md

```diff
@@ -0,0 +1,281 @@
+
+Related discussions:
+* [Support cluster-local Gateways](https://github.com/kubernetes-sigs/gateway-api/discussions/1247)
+* [Scaling Gateway Resources](https://github.com/kubernetes-sigs/gateway-api/discussions/1355)
+* [Manual deployments](https://github.com/kubernetes-sigs/gateway-api/issues/1687)
+* [Merging Gateways](https://github.com/kubernetes-sigs/gateway-api/pull/1863/)
+* [Per-Gateway Infrastructure](https://github.com/kubernetes-sigs/gateway-api/pull/1757)
+
+## Goals
+
+* Provide prescriptive guidance for how in-cluster implementations should behave.
+* Provide requirements (tested by conformance) for how in-cluster implementations should behave.
+
+Note that some changes will be suggestions, while others will be requirements.
+
+## Non-Goals
+
+* Provide guidance to how out-of-cluster implementations should behave. Rather, this document aims to bring consistency between these types.
+
+## Terminology
+
+This document uses a few terms throughout. To ensure consistency, they are defined below:
+
+* In-cluster deployment: refers to an implementation that actuates a `Gateway` by running a data plane in the cluster.
+  This is *often*, but not necessarily, by deploying a `Deployment`/`DaemonSet` and `Service`.
+* Automated deployment: refers to an implementation that automatically deploys the data plane based on a `Gateway`.
+  That is, the user simply creates a `Gateway` resource and the rest is handled behind the scenes by the implementation.
+* Manual deployment: refers to an implementation that does NOT automatically deploy the data plane based on a `Gateway`.
+  This may require, for example, a user to manually create a `Deployment` and `Service` for the data plane, then link this to a `Gateway`.
+
+## Design
+
+This GEP does not introduce new API fields, but rather standardizes how implementations should behave when implementing the existing API.
+
+### Automated Deployments
+
+A simple `Gateway`, as is configured below is assumed to be an automated deployment:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: default
+    port: 80
+    protocol: HTTP
+```
+
+With this configuration, an implementation:
+* MUST mark the Gateway as `Programmed` and provide an address in `Status.Addresses` where the Gateway can be reached on each configured port.
+* MUST label all generated resources (Service, Deployment, etc) with `gateway.networking.k8s.io/metadata.name: my-gateway` (where `my-gateway` is the name of the Gateway resource).
+* MUST provision generated resources in the same namespace as the Gateway if they are namespace scoped resources.
+  * Cluster scoped resources are not recommended.
+* MUST name all generated resources `my-gateway-example` (`<NAME>-<GATEWAY CLASS>`).
```

Comment:
How should this handle multiple resources of the same kind without colliding? Was this intended to be something like "MUST use <pattern> as a name prefix for generated resources"?

We, for example, automatically create multiple Deployments (the proxy instances and a configuration manager) and and multiple Services (admin traffic, proxy traffic, and some utility/status APIs), some associated with one Deployment and some with the others.

We currently do something like the prefix approach. IIRC it's caused some issues because of the name character limit, and I don't think we've found a better way around that than telling users to choose shorter names.

Reply:
Currently it does not. Interesting case. On benefit of deterministic names is a lot of other policies attach via name. For example HPA. So having a common name is pretty handy. That is tricky if we have multiple though...

Reply:
Is this dilemma something we can capture in a TODO section so that we can merge, but don't have to hold up this specific PR?

Reply:
I remain in favor of creating a TODO item for this in the GEP, with the intention that we don't move to `implementable` until we resolve this, but to allow us to move forward with this first large iteration and resolve this separately as its own focus.

Reply:
I think it's going to be very difficult to _require_ this behavior without being disruptive to all in-cluster implementations. I think suggesting a pattern, and stating that it may just serve as a prefix (ie `generateName` instead of `name`) may be sufficient here.